pallets/transfer/src/lib.rs

```diff
@@ -0,0 +1,223 @@
+#![cfg_attr(not(feature = "std"), no_std)]
+
+pub use pallet::*;
+
+#[cfg(test)]
+mod mock;
+
+#[cfg(test)]
+mod tests;
+
+#[frame_support::pallet]
+pub mod pallet {
+    use blake2::digest::{Update, VariableOutput};
+    use blake2::VarBlake2b;
+    use frame_support::{
+        pallet_prelude::*,
+        traits::{Currency, ExistenceRequirement, Vec},
+        unsigned::ValidateUnsigned,
+    };
+    use frame_system::{
+        offchain::{SendTransactionTypes, SubmitTransaction},
+        pallet_prelude::*,
+    };
+    use primitives::*;
+    use sp_core::{ecdsa, H160};
+    use sp_io::{crypto::secp256k1_ecdsa_recover, hashing::keccak_256};
+    use sp_runtime::{traits::UniqueSaturatedInto, SaturatedConversion};
+    use sp_std::str;
+
+    /// Type alias for currency balance.
+    type BalanceOf<T> =
+        <<T as Config>::Currency as Currency<<T as frame_system::Config>::AccountId>>::Balance;
+    type EcdsaSignature = ecdsa::Signature;
+
+    const UNSIGNED_TXS_PRIORITY: u64 = 100;
+
+    #[pallet::config]
+    pub trait Config: SendTransactionTypes<Call<Self>> + frame_system::Config {
+        type Event: From<Event<Self>> + IsType<<Self as frame_system::Config>::Event>;
+        type Currency: Currency<Self::AccountId>;
+        type Call: From<Call<Self>>;
+    }
+
+    #[pallet::pallet]
+    #[pallet::generate_store(pub(super) trait Store)]
+    pub struct Pallet<T>(_);
+
+    // Pallets use events to inform users when important changes are made.
+    // https://substrate.dev/docs/en/knowledgebase/runtime/events
+    #[pallet::event]
+    #[pallet::metadata(T::AccountId = "AccountId")]
+    #[pallet::generate_deposit(pub(super) fn deposit_event)]
+    pub enum Event<T: Config> {
+        TransferToEVM(T::AccountId, H160, Balance),
+        TransferToSubstrate(H160, T::AccountId, Balance),
+    }
+
+    // Errors inform users that something went wrong.
+    #[pallet::error]
+    pub enum Error<T> {
+        /// Use an invalid attestor id.
+        InsufficientBalance,
+        SignatureInvalid,
+        SignatureMismatch,
+        IncorrectNonce,
+    }
+
+    #[pallet::validate_unsigned]
+    impl<T: Config> ValidateUnsigned for Pallet<T> {
+        type Call = Call<T>;
+
+        fn validate_unsigned(_source: TransactionSource, call: &Self::Call) -> TransactionValidity {
+            let valid_tx = |provide| {
+                ValidTransaction::with_tag_prefix("Automata/evm/transfer")
+                    .priority(UNSIGNED_TXS_PRIORITY)
+                    .and_provides([&provide])
+                    .longevity(3)
+                    .propagate(true)
+                    .build()
+            };
+
+            match call {
```

Comment:
We can verify parameters check here, instead of doing check in extrinsics, so we can prevent invalid extrinsics from entering transaction pool, I will do this improvement.

Reply:
done